File: backend/workout-api/src/main/java/tom/projects/monkeylog/service/WorkoutService.java
package tom.projects.monkeylog.service;

import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;
import tom.projects.monkeylog.dto.workout.WorkoutRequest;
import tom.projects.monkeylog.mapper.WorkoutMapper;
import tom.projects.monkeylog.model.workout.Workout;
import tom.projects.monkeylog.repository.workout.WorkoutRepository;
import tom.projects.monkeylog.security.AuthenticatedUser;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;

@Service
@RequiredArgsConstructor
public class WorkoutService {
    private static final String WORKOUT_NOT_FOUND = "Workout not found.";
    private final WorkoutRepository workoutRepository;
    private final WorkoutMapper workoutMapper;

    public List<Workout> all(Workout.Type type) {
        return workoutRepository.findAllByTypeAndUserId(type, AuthenticatedUser.getId());
    }

    public Workout get(Long id) {
        return workoutRepository.findById(id)
                .filter(AuthenticatedUser::isOwner)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, WORKOUT_NOT_FOUND));
    }

    public Workout getActive() {
        return workoutRepository.findAllByTypeAndUserId(Workout.Type.ACTIVE, AuthenticatedUser.getId()).stream()
                .findFirst()
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, WORKOUT_NOT_FOUND));
    }

    public Object count(LocalDate start, LocalDate end) {
        return workoutRepository.findByEndDateBetween(start.atStartOfDay(), end.plusDays(1).atStartOfDay());
    }

    public Workout start() {
        Workout workout = new Workout();
        workout.setName("New empty workout");

        return start(workout);
    }

    public Workout start(Long id) {
        Workout newWorkout = Workout.clone(get(id));

        return start(newWorkout);
    }

    private Workout start(Workout workout) {
        workout.setStartDate(LocalDateTime.now());
        workout.setType(Workout.Type.ACTIVE);
        workout.setUserId(AuthenticatedUser.getId());

        clearActive();

        return workoutRepository.save(workout);
    }

    private void clearActive() {
        workoutRepository
                .deleteAll(workoutRepository.findAllByTypeAndUserId(Workout.Type.ACTIVE, AuthenticatedUser.getId()));
    }

    public Workout complete() {
        Workout workout = workoutRepository.findAllByTypeAndUserId(Workout.Type.ACTIVE, AuthenticatedUser.getId())
                .stream().findFirst()
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, WORKOUT_NOT_FOUND));

        workout.getExerciseGroups().forEach(
                exerciseGroup -> exerciseGroup.getExerciseRows().removeIf(exerciseRow -> !exerciseRow.getIsLifted()));
        workout.getExerciseGroups().removeIf(exerciseGroup -> exerciseGroup.getExerciseRows().isEmpty());
        workout.setEndDate(LocalDateTime.now());
        workout.setType(Workout.Type.COMPLETED);

        return workoutRepository.save(workout);
    }

    public Workout clone(Long id) {
        Workout workout = get(id);
        Workout newWorkout = Workout.clone(workout);
        newWorkout.setName(workout.getName() + " - copy");
        newWorkout.setType(Workout.Type.TEMPLATE);
        newWorkout.setUserId(AuthenticatedUser.getId());

        return workoutRepository.save(newWorkout);
    }

    public Workout save(WorkoutRequest workoutRequest) {
        Workout workout = workoutMapper.workoutRequestToWorkout(workoutRequest);
        workout.setType(Workout.Type.TEMPLATE);
        workout.setUserId(AuthenticatedUser.getId());

        return workoutRepository.save(workout);
    }

    public void delete(Long id) {
        workoutRepository.delete(get(id));
    }
}

File: backend/src/main/kotlin/tom/com/workout/logbook/service/WorkoutService.kt
package tom.com.workout.logbook.service

import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import org.springframework.web.server.ResponseStatusException
import tom.com.workout.logbook.common.dto.IdProjection
import tom.com.workout.logbook.dto.workout.WorkoutCreateRequest
import tom.com.workout.logbook.mapper.toEntity
import tom.com.workout.logbook.model.workout.ExerciseRow
import tom.com.workout.logbook.model.workout.Workout
import tom.com.workout.logbook.model.workout.WorkoutType
import tom.com.workout.logbook.repository.WorkoutRepository
import tom.com.workout.logbook.security.AuthenticatedUser
import java.time.Instant
import java.util.*
import java.util.function.Predicate.not


@Service
class WorkoutService(
    private val workoutRepository: WorkoutRepository,
    private val programService: ProgramService,
) {
    fun getPagedWorkoutIds(workoutType: WorkoutType, pageable: Pageable): Page<IdProjection> {
        return workoutRepository.findAllByWorkoutTypeAndUserId(
            workoutType,
            AuthenticatedUser.id,
            pageable
        )
    }

    fun getWorkouts(ids: List<UUID>): List<tom.com.workout.logbook.model.workout.Workout> {
        return workoutRepository.findAllByIdIn(ids)
    }

    fun getWorkout(id: UUID): tom.com.workout.logbook.model.workout.Workout {
        return workoutRepository.findById(id)
            .filter(AuthenticatedUser::isResourceOwner)
            .orElseThrow { ResponseStatusException(HttpStatus.NOT_FOUND, WORKOUT_NOT_FOUND) }
    }

    fun getActiveWorkouts(): List<tom.com.workout.logbook.model.workout.Workout> {
        return workoutRepository.findAllByWorkoutTypeAndUserId(
            WorkoutType.ACTIVE,
            AuthenticatedUser.id
        );
    }

    fun startWorkout(): tom.com.workout.logbook.model.workout.Workout {
        return tom.com.workout.logbook.model.workout.Workout(name = "New empty workout")
            .run(this::startWorkout)
    }

    fun startWorkout(id: UUID): tom.com.workout.logbook.model.workout.Workout {
        return getWorkout(id).clone().run(this::startWorkout)
    }

    private fun startWorkout(workout: tom.com.workout.logbook.model.workout.Workout): tom.com.workout.logbook.model.workout.Workout {
        workoutRepository.deleteAll(getActiveWorkouts())

        return workout.apply {
            startDate = Instant.now()
            workoutType = WorkoutType.ACTIVE
            userId = AuthenticatedUser.id
        }.let(workoutRepository::save)
    }

    fun complete(): tom.com.workout.logbook.model.workout.Workout {
        val activeWorkout: tom.com.workout.logbook.model.workout.Workout =
            getActiveWorkouts().firstOrNull()
                ?: throw ResponseStatusException(HttpStatus.NOT_FOUND, WORKOUT_NOT_FOUND)

        return activeWorkout.apply {
            endDate = Instant.now()
            workoutType = WorkoutType.COMPLETED
            exerciseGroups.forEach { it.exerciseRows.removeIf(not(ExerciseRow::lifted)) }
            exerciseGroups.removeIf { it.exerciseRows.isEmpty() }
        }.let(workoutRepository::save)
    }

    fun cloneToTemplate(id: UUID) = getWorkout(id).clone().let(workoutRepository::save)

    fun save(workoutRequest: WorkoutCreateRequest): tom.com.workout.logbook.model.workout.Workout {
        return workoutRequest.toEntity().apply {
            workoutType = WorkoutType.TEMPLATE
            userId = AuthenticatedUser.id
            programWeek = workoutRequest.programWeekId?.let(programService::getProgramWeek)
        }.let(workoutRepository::save)
    }

    fun delete(id: UUID) {
        workoutRepository.delete(getWorkout(id))
    }

    companion object {
        const val WORKOUT_NOT_FOUND = "Workout not found."
    }
}